lib/src/rules/void_checks.dart

```diff
@@ -131,10 +133,12 @@ class _Visitor extends SimpleAstVisitor<void> {
     checkedNode ??= node;
     if (expectedType == null || type == null) {
       return;
-    } else if (expectedType.isVoid &&
-            !isTypeAcceptableWhenExpectingVoid(type) ||
-        expectedType.isDartAsyncFutureOr &&
-            (expectedType as InterfaceType).typeArguments.first.isVoid &&
+    }
+
+    var futureOrVoidType =
+        context.typeProvider.futureOrType2(context.typeProvider.voidType);
+    if (expectedType.isVoid && !isTypeAcceptableWhenExpectingVoid(type) ||
+        expectedType == futureOrVoidType &&
             !typeSystem.isAssignableTo(type, _futureDynamicType)) {
```

Comment:
Can you explain the portion of expecting `FutureOr<void>` and `typeSystem.isAssignableTo(type, _futureDynamicType)`?

Reply:
I'm not sure why futureDynamicType was chosen originally. Seems like a strange choice to me. This fix just makes it so that when Null is used where the expected type is `FutureOr<void>`, then the lint rule will not complain "Hey don't assign a Null to `FutureOr<void>`.

If I changed this check to "is the type assignable to Future<void>", for example, then it no longer catches a case like:

```dart
FutureOr<void> a = 1;
```

I think because technically this is legal, `int` is assignable to `FutureOr<void>`, but the lint wishes to call it out as "Don't assign to void."

Reply:
If the return type of `f` in your example were `FutureOr<void>`, I think we want to allow:

* `Future<dynamic>` (tested on [line 58](https://github.com/dart-lang/linter/blob/master/test/rules/void_checks.dart#L58)),
* `Future<void>` (not tested?),
* `Future<Null>` (not tested?),
* `FutureOr<dynamic>` (not tested?),
* `FutureOr<void>` (line 71),
* `FutureOr<Null>` (newly tested in this PR, [line 10](https://github.com/dart-lang/linter/pull/2274/files#diff-e6fb8c48dbd9c0d925e3d68ea6bf3b1b429a4c3f77a105afbb36e3e50dfc07e5))
* `dynamic` (line 72),
* `Null` (line 73),

and nothing else, I think. Right now the language allows, e.g. an int to be returned where `FutureOr<void>` is expected.

Reply:
If we can have this nice exhaustive list of types, I think it would be better to check for it explicitly, like we did for `void`, instead of using a more heavy `isAssignableTo`.

Reply:
Ah ha, good point. I've used isTypeAcceptableWhenExpectingVoid to write a isTypeAcceptableWhenExpectingFutureOrVoid.